pkg/storage/favorite/favorite.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2018-2021 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package favorite
+
+import (
+	user "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+)
+
+// Manager defines an interface for a favorites manager.
+type Manager interface {
+	// ListFavorites returns all resources that were favorited by a user.
+	ListFavorites(userID *user.UserId) ([]*provider.ResourceInfo, error)
+	// SetFavorite marks a resource as favorited by a user.
+	SetFavorite(userID *user.UserId, id *provider.ResourceInfo) error
+	// UnsetFavorite unmarks a resource as favorited by a user.
+	UnsetFavorite(userID *user.UserId, id *provider.ResourceInfo) error
+}
+
+// NewInMemoryManager returns an instance of a favorites manager using an in-memory storage.
+func NewInMemoryManager() Manager {
+	return InMemoryManager{favorites: make(map[string]map[string]*provider.ResourceInfo)}
+}
+
+// InMemoryManager implements the Manager interface to manage favorites using an in-memory storage.
+type InMemoryManager struct {
+	favorites map[string]map[string]*provider.ResourceInfo
+}
+
+func (m InMemoryManager) ListFavorites(userID *user.UserId) ([]*provider.ResourceInfo, error) {
+	favorites := make([]*provider.ResourceInfo, 0, len(m.favorites[userID.OpaqueId]))
```

Comment:
I think the favorites manager should only have a list of ResourceIDs. The ocdav service (and later the gateway) needs to stat (and cache the actual resources). IIRC that is how the shares are handled as well. Hmmm, I can see the point of having the full Resource info in the favorites manager.

Reply:
I see, storing the `ResourceInfo` has the disadvantage that it will eventually be outdated...

I implemented it this way to not have to stat all the resources but maybe a compromise could be to use the storage driver in the favorite manager to be able to more efficiently stat the resources (without the grpc and auth interceptors overhead)